libp2p/connection.nim

```diff
@@ -0,0 +1,120 @@
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import chronos, options, chronicles
+import peerinfo,
+       multiaddress,
+       stream/lpstream,
+       peerinfo,
+       varint,
+       vbuffer
+
+const DefaultReadSize*: uint = 64 * 1024
+
+type
+  Connection* = ref object of LPStream
+    peerInfo*: PeerInfo
+    stream*: LPStream
+    observedAddrs*: Multiaddress
+
+proc newConnection*(stream: LPStream): Connection =
+  ## create a new Connection for the specified async reader/writer
+  new result
+  result.stream = stream
+
+method read*(s: Connection, n = -1): Future[seq[byte]] {.gcsafe.} =
+  result = s.stream.read(n)
+
+method readExactly*(s: Connection,
+                    pbytes: pointer,
+                    nbytes: int): 
+                    Future[void] {.gcsafe.} =
+  result = s.stream.readExactly(pbytes, nbytes)
+
+method readLine*(s: Connection,
+                 limit = 0,
+                 sep = "\r\n"): 
+                 Future[string] {.gcsafe.} =
+  result = s.stream.readLine(limit, sep)
+
+method readOnce*(s: Connection,
+                 pbytes: pointer,
+                 nbytes: int): 
+                 Future[int] {.gcsafe.} =
+  result = s.stream.readOnce(pbytes, nbytes)
+
+method readUntil*(s: Connection,
+                  pbytes: pointer,
+                  nbytes: int,
+                  sep: seq[byte]): 
+                  Future[int] {.gcsafe.} =
+  result = s.stream.readUntil(pbytes, nbytes, sep)
+
+method write*(s: Connection, 
+              pbytes: pointer, 
+              nbytes: int): 
+              Future[void] {.gcsafe.} =
+  result = s.stream.write(pbytes, nbytes)
+
+method write*(s: Connection, 
+              msg: string, 
+              msglen = -1): 
+              Future[void] {.gcsafe.} =
+  result = s.stream.write(msg, msglen)
+
+method write*(s: Connection, 
+              msg: seq[byte], 
+              msglen = -1): 
+              Future[void] {.gcsafe.} =
+  result = s.stream.write(msg, msglen)
+
+method close*(s: Connection) {.async, gcsafe.} =
+  await s.stream.close()
+  s.closed = true
+
+proc readLp*(s: Connection): Future[seq[byte]] {.async, gcsafe.} =
+  ## read lenght prefixed msg
+  var
+    size: uint
+    length: int
+    res: VarintStatus
+  var buffer = newSeq[byte](10)
+  try:
+    for i in 0..<len(buffer):
+      await s.readExactly(addr buffer[i], 1)
+      res = LP.getUVarint(buffer.toOpenArray(0, i), length, size)
+      if res == VarintStatus.Success:
+        break
+    if res != VarintStatus.Success or size > DefaultReadSize:
+      buffer.setLen(0)
+      result = buffer
+      return
+    buffer.setLen(size)
+    if size > 0.uint:
+      await s.readExactly(addr buffer[0], int(size))
+  except LPStreamIncompleteError, LPStreamReadError:
```

Comment:
same here then.. why are these two errors handled differently than say.. a `TransportIncompleteError` or whatever else the underlying transport might raise? 
Basically, it seems that the `readUntil` and `readExactly` methods of chronosstream rename different exceptions (`TransportIncompleteException`?) - is this the full set of non-defect exceptions that _can_ be raised by all chronos transports when calling their readExactly? why is `readExactly` and `readUntil` different - judging from their name, the only difference is predicate upon which they stop reading (back when they were added to chronos, there was an argument that you should simply have a single `readWhile` taking a predicate)?  or is the behaviour supposed to be different depending on which exception leaks from the underlying stream implementation?

Reply:
So there are a few of questions here.

> why are these two errors handled differently than say.. a TransportIncompleteError or whatever else the underlying transport might raise?

`TransportIncompleteError` gets remapped to an `AsyncStreamIncompleteError`, which I then remap to an `LPStreamIncompleteError` that get raised by all `LPStream` descendants. I did this because I want to keep `LPStream` independent from chronos streams. Tho it's API was borrowed from chronos streams, the relation ends there, they are different beasts intended for different purposes. The idea behind `LPStream` is/was that I can swap chronos streams for something else down the road (not that I'm planning on doing that, but this model allows for that). 

> is this the full set of non-defect exceptions that can be raised by all chronos transports when calling their readExactly?

Yes

> why is readExactly and readUntil different - judging from their name, the only difference is predicate upon which they stop reading (back when they were added to chronos, there was an argument that you should simply have a single readWhile taking a predicate)?

I ask the same question, they are pretty similar and should probably be unified. I left it out for now due to time constraints.

This all should be revisited when we do a re-evaluation of the streaming interface we want to stick to.

Reply:
> TransportIncompleteError gets remapped to an AsyncStreamIncompleteError

where? I see it here: https://github.com/status-im/nim-libp2p/pull/32/files#diff-7a0098b7313dad13f4cafc1a3d4febadR65

but not: https://github.com/status-im/nim-libp2p/pull/32/files#diff-7a0098b7313dad13f4cafc1a3d4febadR42

from what can tell it's the latter being called. is there more remapping going on somewhere deep in chronos that you're relying on?

Reply:
The transport can also raise `TransportError` and a bunch of other ones - those should bubble up?

Reply:
Yes, basically the transport exceptions get remapped to AsyncStream exceptions in chronos - https://github.com/status-im/nim-chronos/blob/master/chronos/streams/asyncstream.nim#L274

In general, I only care about the AsyncStream exceptions, all transport exceptions that aren't handled/re-maped by chronos async stream should bubble up. That Transport exception in https://github.com/status-im/nim-libp2p/pull/32/files#diff-7a0098b7313dad13f4cafc1a3d4febadR65,  is a mistake and needs to be removed.

Reply:
uh, ok. that seems pretty hard to reason about - when you call this method, you're expected to:
```
try:
  let buf = s.readLp()
  if buf.len == 0:
    # handle varint decoding errors by dropping peer, most likely
  else:
    ...
  except (not Defect):
    # handle other errors by dropping peer, most likely (unless it was a close error, in which case you need more special code)
```

two problem handling mechanisms? confusing.

Reply:
Not really, anything that `readLp()` throws in unrecoverable, everything else is handled inside it. You might want to intercept it for logging purposes, but should not try to recover from it. 

If you want to gracefully shutdown in this type of scenarios, there are/should be global exception hooks/supervisors (does nim has this?) that would catch the exception, log it and then exit in a graceful manner. 

If you have a long running process and want to keep it running, you should allow the process exit and then have `init` or some other monitoring/supervising software restart it, trying to recover from within the application is almost always a bad idea.

Reply:
> If you want to gracefully shutdown in this type of scenarios, there are/should be global exception hooks/supervisors (does nim has this?) that would catch the exception, log it and then exit in a graceful manner.

BTW, I should mention that this is something that the consuming application should decide on, not the library.

Reply:
wait what? `transporterror` is something that is raised when the remote peer closes the connection - at least from what I can tell from a casual reading - that doesn't seem very.. unusual, and is semantically no different from the peer sending too few varint bytes or whatever - it needs to be handled (by deregistering the peer, and that's it)

Reply:
Sure, that's the bubble up part, but I don't discriminate which error that is, I simply log and disconnect the peer or exit the loop/etc... In this case it makes sense and I treat all errors equally at that level. Also, I should mention that when I say unrecoverable in the context of a connection, it's different from unrecoverable for the whole application - I should have been more explicit about it. 

The argument still stands tho, I don't try to recover from a transport error, in fact I don't even care whether it's a transport error or something else, I simply know to terminate/close the connection and move on.

Reply:
but how is a disconnect different from a peer sending you invalid data? would you try to recover from someone sending you bad varints or incomplete streams? how? you have no idea what state they're in so the best you can do is disconnect them and start over

Reply:
Also, your snippet isn't confusing at all, it's the standard way of handling exceptional conditions.

```
try:
  let buf = s.readLp()
  if buf.len == 0:
    # handle varint decoding errors by dropping peer, most likely
  else:
    ...
except (not Defect):
    # handle other errors by dropping peer, most likely (unless it was a close error, in which case you need more special code)
```

I'm failing to see where the perceived increase in cognitive load is creeping in to this otherwise perfectly looking snippet of beautiful exception aware code :)?

Reply:
> I'm failing to see where the perceived increase in cognitive load is creeping in to this otherwise perfectly looking snippet of beautiful exception aware code :)?

never mind that you both have to handle exceptions and know that the magic empty buffer means a special category of errors from which you recover exactly the same way as the exceptions, so you can't _just assume_ the buffer contains valid stuff when you use it. yup, makes perfect sense - gotta exercise those all those fancy keywords or you'll forget that they exist ;)
also, the peer disconnecting is exceptional, while it sending bad data is normal. yup, perfectly sensible ;)

Reply:
another way to phrase that: if I'm going down the exception path, at least I'd want to reap the benefits of being able to ignore any kind of errors and have it return valid data only so that the calling code can assume that the primary intended purpose of the api is fulfilled on return, always - in this case, getting bytes as the name suggests.

Reply:
> but how is a disconnect different from a peer sending you invalid data? would you try to recover from someone sending you bad varints or incomplete streams? how? you have no idea what state they're in so the best you can do is disconnect them and start over

Ah, good questions. 

Sending bad data is a logical error, a disconnect is an unrecoverable error no matter how you try to handle it. With logical errors it's up to the specific layer to decide what to do with it. However, with length prefixed data, it's impossible to send more data than what is specified as the length of the message, so you either "hang" waiting for the full message to be arrive, or you get a partial message because the length was incorrect. In the latter case, you'd probably find out once you try to do something with the message, which might or might not happen at the same level the message was read. Bottom line, length prefixed messages are there specifically so that you don't have to fiddle with weird conditions like that - i.e did I read all of it or did the connection simply dropped, etc... And if the message is read correctly, it's up to the application to decide whether it's correct or not and handle the connection accordingly.

In the case of bad varint, currently I simply return an empty buffer. Perhaps I should throw as well - probably yes.

Reply:
> In the case of bad varint, currently I simply return an empty buffer. Perhaps I should throw as well - probably yes.

To clarify, I'm not treating bad varint as an exceptional condition, so I don't disconnect on it. If I do (which I most definitely should, now that I think about it), I would throw and it would be handled same as other non recoverable exceptions.

Reply:
Just to clarify my position so that it is clear where I'm coming from. To me exceptions are still a concise and elegant way of doing structured error handling. That doesn't mean that they are perfect - everything has corner cases (and respective workarounds). That doesn't mean however (at least to me), that it is a flawed model and it should be disregarded completely.

Now, I do agree that there are objective scenarios where they aren't the right tool for the job or can't be used at all - for example, lack of runtime support, embedded systems, etc... 

With that said, exceptions have well defined patterns and best practices that when applied correctly provide reasonable guarantees of correctness.

I would love to see what makes the above snippet wrong or incorrect; or how it can be made better. :)

Reply:
> never mind that you both have to handle exceptions and know that the magic empty buffer means a special category of errors from which you recover exactly the same way as the exceptions, so you can't just assume the buffer contains valid stuff when you use it. yup, makes perfect sense 

Well, in the case of a bad varint, it's definitely an exceptional condition and I should throw (fixing as I write this). 

But, how do you know what's bad data and what isn't?

Reply:
> or is the behaviour of this api supposed to be different depending on which exception leaks from the underlying stream implementation (exception vs 0-length buffer return)?

I failed to address this part. With length prefixed data, there should never be an empty message/buffer, thats why returning on 0 buffer is plain wrong (thanks for pointing it out!). Bot disconnects and malformed varints are equally bad and should go through the same exception handling path - i.e. cleanup the connection. 

With logical errors (malformed protobuf, invalid flags, etc...) it is up to the application to decide what to do with it. It's worth pointing out that libp2p messages are onion like, they could potentially contain many, length prefixed wrapped messages, but each level only cares about it's own layer, for example, `mplex` reads the header and decodes what channel the message belongs to and then routs it there, it doesn't care what the data field contains, or weather it's addressed to a core libp2p protocol like identify or a user provided one - that's someone else's job (multistream-select). The next layer might decide that the message is malformed or missing data and handle it accordingly. But the distinction stands, truly unrecoverable errors and logical/application errors. For example, if a user defined handler got a malformed protobuf/messagepack/byson etc message, it can decide to disconnect, re-request it or do whatever else it wants with it, at that point I've handed it over to it and as far as libp2p is concerned the message was perfectly fine because it contained the correct amount of bytes and the stream did not die.

Reply:
well, since we're here, I'm addressing this from a internal consistency perspective mainly, not exceptions-vs-something-else. 

If a function is called `readLp`, there are two possible behaviours I'd expect from it by just looking at the name: a) return the data including the length prefix and b) return the data without the prefix. anything else, I'd expect to be handled via an error handling notification mechanism of choice - in this case exceptions.

depending on circumstances, a) or b) might be the right thing, but all things equal, I'd go for b) as `readLp` would de-onion the length prefix and give me next layer - in particular, this would mean that zero-length seq is actually a valid return value (and the next layer can indeed figure out what that means).

now - this morphed into a much longer, and quite unrelated, discussion about the raison-d'etre for exceptions - notably though, at the start it wasn't even about exceptions or not, just inconsistent semantics (just like the buffer reset, that smelled bad). in particular, why are some exceptions magic and remapped, and others no? there was no pattern in the code, as it looked at the start, because exceptions different abstraction layers were being remapped randomly in similar functions. it looks quite odd that it remaps some exceptions but not others, to begin with, but untangling that is beyond a casual code-smell review like this ;) trying to actually understand control flows in those cases is.. hard because you have to read implementations of _all_ dependent code instead of stopping at the signature.

the smell that made me even look at it was indeed that two error handling mechanisms were used in the same api. there are exceptional (haha) cases where that makes sense, but rarely in exception-heavy api like this - in particular, it happens when the api is designed ground up to use exceptions only in exceptional cases - in those cases where panics/aborts would be used in other languages (defects, as they're called in nim) - but this definitely looks like exception maxi code so I'm reviewing it from that perspective ;)

Reply:
A small remark about nomenclature:

We usually call an error "unrecoverable" if it's something that should never arise in a correctly implemented program. It's a violation of a pre-condition of a proc or another invariant, in other words - a bug. 

Recoverable errors, on the other hand, represent situations that you have anticipated and for which you can have a discussion about how the software should behave (e.g. if another peer sends us invalid data, our policy is to disconnect them and reduce their reputation). 

In this sense, the exceptions discussed here are recoverable (they should derive from `CatchableError`), but I understand why you call them non-recoverable and why you find it easier to propagate them implicitly through the libp2p layers and handle them in a single place in the outermost layer.

Jacek's main concern about exceptions is that they make it easier for the programmer to not spend any time thinking about errors when using a certain API. Simple refactorings can break previously baked-in invisible assumptions in the code. I hope that in the future, we'll improve Nim's exception tracking capabilities enough to make sure that using exceptions is just as robust as using error types.

Reply:
> well, since we're here, I'm addressing this from a internal consistency perspective mainly, not exceptions-vs-something-else.

Oh! Then I misread the intent. Thankfully we still raised valid points and uncovered some inconsistencies in the implementation 👍 

> If a function is called `readLp`, there are two possible behaviours I'd expect from it by just looking at the name: a) return the data including the length prefix and b) return the data without the prefix. anything else, I'd expect to be handled via an error handling notification mechanism of choice - in this case exceptions.

> depending on circumstances, a) or b) might be the right thing, but all things equal, I'd go for b) as `readLp` would de-onion the length prefix and give me next layer -
Currently `readLp` implements option b, it reads the length prefix and then reads the required bytes from the wire, giving back only the contents of the message, without the length prefix.

> in particular, this would mean that zero-length seq is actually a valid return value (and the next layer can indeed figure out what that means).

Agree on this - In particular, if the message contains a well encoded 0 length prefix and no data, it is a valid message for all intents and purposes. That said, a malformed varint, is the exceptional condition that needs to be handled according to the unrecoverable exception rule (I'll stick to my initial definition for now @zah).

>  in particular, why are some exceptions magic and remapped, and others no

This is a great point. IMO this is a consequence of using wrong abstractions at the lower levels. I don't want to go too deeply into this as this is a larger discussion around chronos and streams in general. But I agree on the inconsistency and struggled with it myself when writing my implementation - not happy about it at all.